apps/vmq_server/src/vmq_mqtt_fsm.erl

```diff
@@ -315,6 +315,9 @@ connected(#mqtt_pubrec{message_id=MessageId}, State) ->
                retry_queue=set_retry(pubrel, MessageId, RetryInterval, RetryQueue),
                waiting_acks=maps:update(MessageId, PubRelFrame, WAcks)},
             [PubRelFrame]};
+        #mqtt_pubrel{} ->
+            lager:debug("Duplicate pubrec ~p ,ignore", [MessageId]),
+            {State,[]};
```

Comment:
I've been looking into thee issue a bit further and think I understand it - it does seem we don't handle it correctly when the client resends a pubrec, but we can't ignore it, the spec says this (ref [MQTT-4.3.3-1] in docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html):

```
MUST send a PUBREL packet when it receives a PUBREC packet from the receiver. This PUBREL packet MUST contain the same Packet Identifier as the original PUBLISH packet.
```

So I think you're on the right track but I think the correct behaviour would be to do something like:

```erlang
        #mqtt_pubrel{} = Frame ->
            {State,[Frame]};
```

Which would cause the pubrel to be re-sent. I'll write up a test reproducing this on MQTT 3.1.1 and MQTT 5.0 so we can never break this again.

Reply:
We probably also need to fix this in the MQTT 5.0 fsm

Reply:
/cc @dergraf - just FYI